light-client/src/operations/voting_power.rs

```diff
@@ -220,75 +220,181 @@ fn non_absent_vote(
+            &light_block.signed_header,
+            &light_block.validators,
+            trust_threshold,
+        );
+
+        // ensure the result is "Ok"
+        assert!(result_ok.is_ok());
+
+        // ensure the result matches the expected result
+        assert_eq!(result_ok.ok().unwrap(), EXPECTED_RESULT);
     }
 
-    fn read_test_case(file_path: impl AsRef<Path>) -> TestCase {
-        serde_json::from_str(read_json_fixture(file_path).as_str()).unwrap()
+    #[test]
+    fn test_all_signatures_nil() {
+        let vp_calculator = ProdVotingPowerCalculator::default();
+        let trust_threshold = TrustThreshold::default();
+
+        let validator_set = ValidatorSet::new(vec!["a", "b"]);
+        let vals = validator_set.clone().validators.unwrap();
+        let header = Header::new(&vals);
+        let votes = vec![
+            TestgenVote::new(vals[0].clone(), header.clone()).is_nil(true),
+            TestgenVote::new(vals[1].clone(), header.clone()).is_nil(true),
+        ];
+        let commit = Commit::new_with_votes(header.clone(), 1, votes);
+        let signed_header = generate_signed_header(&header, &commit).unwrap();
+        let valset = validator_set.generate().unwrap();
+
+        let result_ok = vp_calculator.voting_power_in(&signed_header, &valset, trust_threshold);
+
+        // ensure the result is "Ok"
+        assert!(result_ok.is_ok());
+
+        // ensure the result matches the expected result
+        assert_eq!(result_ok.ok().unwrap(), EXPECTED_RESULT);
     }
 
-    fn run_all_tests() {
-        for entry in fs::read_dir(TEST_FILES_PATH).unwrap() {
-            let entry = entry.unwrap();
-            let tc = read_test_case(entry.path());
-            let name = entry.file_name().to_string_lossy().to_string();
-            run_test(tc, name);
-        }
+    #[test]
+    fn test_one_invalid_signature() {
+        let vp_calculator = ProdVotingPowerCalculator::default();
+        let trust_threshold = TrustThreshold::default();
+
+        let mut testgen_lb = TestgenLightBlock::new_default(10);
+        let mut commit = testgen_lb.commit.clone().unwrap();
+        let mut votes = commit.votes.unwrap();
+        let vote = votes.pop().unwrap();
+        let header = vote.clone().header.unwrap().chain_id("bad-chain");
+        votes.push(vote.header(header));
+
+        commit.votes = Some(votes);
+        testgen_lb.commit = Some(commit);
+        let light_block: LightBlock = testgen_lb.generate().unwrap().into();
+
+        let result_err = vp_calculator.voting_power_in(
+            &light_block.signed_header,
+            &light_block.validators,
+            trust_threshold,
+        );
+
+        // ensure the result is "Err"
+        assert!(result_err.is_err());
+
+        // ensure the result produces an InvalidSignature error
+        let err_str = format!("{:?}", result_err.err().unwrap());
+        assert!(err_str.contains("InvalidSignature"));
```

Comment:
The trouble with using this approach to checking the error is, what if the name of the error changes? Most IDEs won't automatically refactor the error name in the string (unless you do a simple search-and-replace operation, which is usually problematic). A better way would be to do something like the following:

```suggestion
        let err = result_err.err().unwrap();
        match err {
            VerificationError::InvalidSignature { .. } => {}
            _ => panic!("unexpected error: {:?}", err),
        }
```

Reply:
You're right! Thanks @thanethomson :)